opm/simulators/wells/MSWellHelpers.cpp

```diff
@@ -220,46 +220,35 @@ invDX(const MatrixType& D, VectorType x, DeferredLogger& deferred_logger)
 }
 
 template <typename ValueType>
-ValueType calculateFrictionFactor(const double area, const double diameter,
-                                  const ValueType& w, const double roughness,
-                                  const ValueType& mu)
+ValueType frictionPressureLoss(const double l, const double diameter,
+                               const double area, const double roughness,
+                               const ValueType& density,
+                               const ValueType& w, const ValueType& mu)
 {
-
-    ValueType f = 0.;
     // Reynolds number
     const ValueType re = abs( diameter * w / (area * mu));
 
-    if ( re == 0.0 ) {
-        // make sure it is because the mass rate is zero
-        assert(w == 0.);
-        return 0.0;
-    }
+    constexpr double re_value1 = 2000.;
+    constexpr double re_value2 = 4000.;
 
-    const ValueType re_value1 = 2000.;
-    const ValueType re_value2 = 4000.;
+    ValueType result;
 
     if (re < re_value1) {
-        f = 16. / re;
-    } else if (re > re_value2){
-        f = haalandFormular(re, diameter, roughness);
-    } else { // in between
-        const ValueType f1 = 16. / re_value1;
-        const ValueType f2 = haalandFormular(re_value2, diameter, roughness);
-
-        f = (f2 - f1) / (re_value2 - re_value1) * (re - re_value1) + f1;
+        // not using the formula directly because of the division with w might cause singular Jacobian matrix
+        result = 32.* mu * l * abs(w) / (area * diameter *diameter * density);
+    } else {
+        ValueType f;
+        if (re > re_value2){
+            f = haalandFormular(re, diameter, roughness);
+        } else { // in between
+            const ValueType f1 = 16. / re_value1;
+            const ValueType f2 = haalandFormular(re_value2, diameter, roughness);
+            f = (f2 - f1) / (re_value2 - re_value1) * (re - re_value1) + f1;
+        }
+        // \Note: a factor of 2 needs to be here based on the dimensional analysis
+        result = 2. * f * l * w * w / (area * area * diameter * density);
     }
-    return f;
-}
-
-template <typename ValueType>
-ValueType frictionPressureLoss(const double l, const double diameter,
-                               const double area, const double roughness,
-                               const ValueType& density,
-                               const ValueType& w, const ValueType& mu)
-{
-    const ValueType f = calculateFrictionFactor(area, diameter, w, roughness, mu);
-    // \Note: a factor of 2 needs to be here based on the dimensional analysis
-    return 2. * f * l * w * w / (area * area * diameter * density);
+    return result;
```

Comment:
One slightly alternative way of writing this is to use early return.  That **might** be easier to read since the number of nested conditionals is smaller
```C++
if (re < re_value1) {
    return 32 * mu * l * abs(w) / (area * diameter *diameter * density);
}

const auto f = (re > re_value2)
    ? haalandFormular(re, diameter, roughness)
    : intermediateReynoldsRange(...);

return 2 * f * l * w * w / (area * area * diameter * density);
```
If we don't want to introduce the `intermediateReynoldsRange()` then we'll still need one level of nested conditionals.

Reply:
There have been a little mixed opinions about early return, mostly related to how people do debugging.  I do see the advantage of reducing one less level of conditions. 

And let us not introduce extra function for here. I will update the PR.